pkg/store/postings_codec.go

```diff
@@ -58,6 +57,67 @@ func isDiffVarintSnappyStreamedEncodedPostings(input []byte) bool {
 	return bytes.HasPrefix(input, []byte(codecHeaderStreamedSnappy))
 }
 
+var (
+	readersPool = sync.Pool{
+		New: func() interface{} {
+			return s2.NewReader(nil, s2.ReaderMaxBlockSize(4<<10))
```

Comment:
I thought the other day that perhaps we could not specify a `New` function and for `gRPC` we could use the default block size whereas for postings we could use the average length of the last 10 inputs, for example. What do you think?

Reply:
The problem is that reader max block size needs to be larger than writer block size; so if the average goes down we cannot decode anymore with that approach i think ( i might be wrong )? Using the minimum 4kb feels more traceable to me somehow.

Reply:
This change still feels backwards incompatible since i could have cached postings somewhere that were written with larger 64kb block size which might be problematic.

Reply:
I think the s2 reader ensures that the buffer is enlarged if needed https://github.com/klauspost/compress/blob/master/s2/reader.go#L152 so the only downside AFAICT is that there could be 1 extra allocation. The only complication I see here is that ideally there would be a way to access the capacity of the internal buffer so that we could pre-allocated a more appropriately sized buffer.

Reply:
https://github.com/klauspost/compress/pull/832 trying to push something forward here to allow us better estimate how big the buffer should be

Reply:
Oh, at least in tests i got issues when i used a smaller ReaderBlockSize. It would return no postings.

Reply:
Ok, in the latest change i added some capacity tracking to the reader pool and use the running average of the capacity of the 10 last returned readers as an estimate for new readers.

Reply:
Had to revert my changes again, if we dont bound max block sizes the readers will just grow until 64kb again, which would suffer the same issues eventually. I think we can only bound the block sizes.

Reply:
Or is it because the benchmarks load up a lot of postings whereas the problematic case is about small postings lists?

Reply:
That might be the issue with the benchmarks but we dont limit the block size it might easily happen in someones prod too; right?

Reply:
I think it doesn't matter because the other choice is allocating []byte for the whole decoded response which in this case will be much bigger than https://github.com/google/snappy/blob/main/framing_format.txt#L106. If suddenly smaller postings come in then it's also OK to reuse the readers with bigger reader buffers because they will be freed eventually.